File: 1_stack/单调栈/力扣加加/移除或保留字符/321. 拼接最大数.ts
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @param {number} k
 * @return {number[]}
 * 现在从这两个数组中选出 k (k <= m + n) 个数字拼接成一个新的数，要求从同一个数组中取出的数字保持其在原数组中的相对顺序。
 */
const maxNumber = (nums1: number[], nums2: number[], k: number): number[] => {
  let res: number[] = []

  for (let i = 0; i <= k; i++) {
    if (i <= nums1.length && k - i <= nums2.length) {
      const cand = merge(pickMaxK(nums1, i), pickMaxK(nums2, k - i))
      // 数组比大小隐式转换成数字(parseInt)
      if (cand > res) res = cand
    }
  }

  return res

  /**
   * @description  见402. 移掉 K 位数字
   * @param arr 待选数组
   * @param k 顺序选出K个数保证组成的数字最大
   */
  function pickMaxK(arr: number[], k: number) {
    let drop = arr.length - k
    const stack: number[] = []
    for (let i = 0; i < arr.length; i++) {
      while (drop > 0 && stack.length > 0 && stack[stack.length - 1] < arr[i]) {
        stack.pop()
        drop--
      }
      stack.push(arr[i])
    }
    // 注意slice
    return stack.slice(0, k)
  }

  /**
   *
   * @param arr1 归并排序的数组
   * @param arr2 归并排序的数组
   */
  function merge(nums1: number[], nums2: number[]) {
    // 一直shift 先转成deque比较好
    const res: number[] = []

    while (nums1.length || nums2.length) {
      // 数组比大小隐式转换成数字(parseInt)
      const bigger = nums1 > nums2 ? nums1 : nums2
      res.push(bigger.shift()!)
    }
    return res
  }
}

// console.log(maxNumber([3, 4, 6, 5], [9, 1, 2, 5, 8, 3], 5))
console.log(maxNumber([6, 7, 5], [4, 8, 1], 3))

File: 22_专题/字典序列删除/321. 拼接最大数.py
"""
给定长度分别为 m 和 n 的两个数组，其元素由 0-9 构成
现在从这两个数组中选出 k (k <= m + n) 个数字拼接成一个新的数，
要求从同一个数组中取出的数字保持其在原数组中的相对顺序。
求满足该条件的最大数。结果返回一个表示该最大数的长度为 k 的数组。
"""

from collections import deque
from typing import List


class Solution:
    def maxNumber(self, nums1: List[int], nums2: List[int], k: int) -> List[int]:
        def pickMax(nums: List[int], k: int) -> List[int]:
            """选k个数拼成最大数  栈顶肯定要最大 单减的单调栈"""
            stack = []
            drop = len(nums) - k
            for num in nums:
                while drop and stack and stack[-1] < num:
                    stack.pop()
                    drop -= 1
                stack.append(num)
            return stack[:k]

        def merge(A: List[int], B: List[int]) -> List[int]:
            """合并两个数组，字典序最大"""
            res = []
            nums1, nums2 = deque(A), deque(B)
            while nums1 or nums2:
                bigger = nums1 if nums1 > nums2 else nums2
                res.append(bigger.popleft())
            return res

        return max(
            merge(pickMax(nums1, i), pickMax(nums2, k - i))
            for i in range(k + 1)
            if i <= len(nums1) and k - i <= len(nums2)
        )
